### PR TITLE
Optimize str::repeat

### PR DESCRIPTION
**Pull request by sinkuu**
Improves the performance of `str::repeat` by bulk copying. Here is the benchmarks of `"abcde".repeat(n)`:

|`n`|old [ns/iter]|new [ns/iter]|diff [%]|
---|---|---|---
|1|27.205|27.421|+0.794|
|2|27.500|27.516|+0.0581|
|3|27.923|27.648|-0.985|
|4|31.206|30.145|-3.40|
|5|35.144|31.861|-9.34|
|7|43.131|34.621|-19.7|
|10|54.945|36.203|-34.1|
|100|428.31|52.895|-87.7|

### CODE REVIEW

File: src/liballoc/str.rs [hunk 1]
```diff
@@ -2066,9 +2067,39 @@ impl str {
...
     #[stable(feature = "repeat_str", since = "1.16.0")]
     pub fn repeat(&self, n: usize) -> String {
-        let mut s = String::with_capacity(self.len() * n);
-        s.extend((0..n).map(|_| self));
-        s
+        if n == 0 {
+            return String::new();
+        }
+
+        // n = 2^k + l (2^k > l)
```
Comment (rkruppe) [inline src/liballoc/str.rs]: Nit: `l` looks like `1`. Better use another letter.

Comment (sanmai-NL) [reply src/liballoc/str.rs]: Better not use single letter identifiers at all...

**Comment by Centril** [comment]
Since this introduces new unsafe bits, and introduces non-trivial logic, I think more comments about why this is safe would be good as well as an overall explanation of the logic.

**Comment by dtolnay** [comment]
Nice work! I tried some tests like the following on my computer and see numbers similar to yours.

```rust
#[bench]
fn repeat100(b: &mut test::Bencher) {
    b.iter(|| { S.repeat(100); });
}
```

|`n`|old [ns/iter]|new [ns/iter]|diff [%]|
---|---|---|---
1 | 28 | 28 | 0
2 | 28 | 28 | 0
3 | 29 | 29 | 0
4 | 33 | 31 | −6.1
5 | 37 | 34 | −8.1
7 | 46 | 35 | −24.
10 | 59 | 37 | −37.
100 | 444 | 65 | −85.

**Pull request merged**